### PR TITLE
Native exception handling

### PR DESCRIPTION
**Pull request by equinox2k**
As per...

https://github.com/mono/Embeddinator-4000/issues/484

I have implemented Rolph's feedback for Native exception handling for apple devices and seems to work great.

I have added a command line param ne/nativeexception which will make objcgen use the technique he mentioned

### CODE REVIEW

File: support/mono_embeddinator.c [hunk 1]
```diff
@@ -292,12 +292,16 @@ MonoMethod* mono_embeddinator_lookup_method(const char* method_name, MonoClass *
 
 void mono_embeddinator_throw_exception(MonoObject *exception)
 {
+    #if defined(__OBJC__) && defined(NATIVEEXCEPTION)
```
Comment (rolfbjarne) [inline support/mono_embeddinator.c]: Style: conditional compilation code is not indented.

File: objcgen/embedder.cs [hunk 2]
```diff
@@ -678,6 +692,8 @@ public int Compile ()
 							mtouch.Append ($"--assembly-build-target=@all=framework={LibraryName}.framework ");
 							mtouch.Append ($"--target-framework {GetTargetFramework ()} ");
 							mtouch.Append ($"\"--gcc_flags=-force_load {Path.GetFullPath (sdk_output_file)}\" ");
+							if (NativeException)
+								mtouch.Append("--marshal-managed-exceptions=throwobjectivecexception ");
```
Comment (rolfbjarne) [inline objcgen/embedder.cs]: This should be done for mmp as well.

Also style: space before opening parenthesis in method calls.

File: objcgen/embedder.cs [hunk 3]
```diff
@@ -290,7 +292,7 @@ public int Compile ()
 					break;
 				case Platform.iOS:
 					build_infos = new BuildInfo[] {
-					new BuildInfo { Sdk = "iPhoneOS", Architectures = new string [] { "armv7", "armv7s", "arm64" }, SdkName = "iphoneos", MinVersion = "8.0", XamariniOSSDK = "MonoTouch.iphoneos.sdk" },
+					new BuildInfo { Sdk = "iPhoneOS", Architectures = new string [] { "armv7", "armv7s", "arm64" }, SdkName = "iphoneos", MinVersion = "8.0", XamariniOSSDK = "MonoTouch.iphoneos.sdk", CompilerFlags = "-fembed-bitcode", LinkerFlags = "-fembed-bitcode" },
```
Comment (rolfbjarne) [inline objcgen/embedder.cs]: Why is this needed? It looks unrelated.

File: objcgen/embedder.cs [hunk 4]
```diff
@@ -629,7 +642,8 @@ public int Compile ()
 								mmp.Append ("--debug ");
 							mmp.Append ("-p "); // generate a plist
 							mmp.Append ($"--target-framework {GetTargetFramework ()} ");
-							mmp.Append ($"\"--link_flags=-force_load {Path.GetFullPath (sdk_output_file)}\" ");
+							string extensionFlag = Extension ? "-fapplication-extension" : "";
```
Comment (rolfbjarne) [inline objcgen/embedder.cs]: Why is this needed? It looks unrelated.

File: objcgen/embedder.cs [hunk 5]
```diff
@@ -388,6 +390,17 @@ public int Compile ()
 							common_options.Append ("-DTOKENLOOKUP ");
 						}
 					}
+
+					if (Extension)
+					{
+						common_options.Append("-fapplication-extension ");
```
Comment (rolfbjarne) [inline objcgen/embedder.cs]: Style 1: single line if statements do not have braces (or if they do, the opening brace is at the end of the if line).
Style 2: space before opening parenthesis in method calls.

File: objcgen/driver.cs [hunk 6]
```diff
@@ -59,6 +59,7 @@ public static int Main2 (params string [] args)
 			var os = new OptionSet {
 				{ "c|compile", "Compiles the generated output", v => embedder.CompileCode = true },
 				{ "e|extension", "Compiles the generated output as extension safe api", v => embedder.Extension = true },
+				{ "ne|nativeexception", "Compiles the generated output to throw native exceptions (Apple only)", v => embedder.NativeException = true },
```
Comment (rolfbjarne) [inline objcgen/driver.cs]: I'm not sure `ne` works correctly (would it interfere with an -n and an -e option?) I'd just go for the full version only (`"nativeexception", "Compiles the ..."`)

**Comment by lemonmojo** [comment]
Does this allow access to the exception type and other properties from ObjC?

File: objcgen/embedder.cs (hunk 3 above)
Comment (chamons) [reply objcgen/embedder.cs]: Default "new" xcode iOS project are defaulting bitcode to on, so it's likely an attempt to shut that up.

Comment (equinox2k) [reply objcgen/embedder.cs]: looks like my other pull request has got merged into this pull

File: objcgen/embedder.cs (hunk 4 above)
Comment (equinox2k) [reply objcgen/embedder.cs]: looks like my other pull request has got merged into this pull

Comment (chamons) [reply objcgen/embedder.cs]: You can just add another commit that reverts these bits and we'll squash it away on land.

**Comment by equinox2k** [comment]
@rolfbjarne Ive made appropriate changed and should be good now...

**Comment by chamons** [comment]
@lemonmojo This PR passes ```marshal-managed-exceptions=throwobjectivecexception``` to mmp/mtouch which will convert managed exceptions into objc ones that you can catch/handle in ObjC as you wish.

**Comment by chamons** [comment]
I agree with @rolfbjarne  that this needs tests, but am pleased with the PR overall. I've filed a tracking issues to note the lack of tests, and this can go in now.

https://github.com/mono/Embeddinator-4000/issues/633

If you want to tack a stab, I'm happy to point you in the right direction, else I'll bang one out early next week.

**Comment by lemonmojo** [comment]
@chamons I understand that but since the managed exceptions will be converted into NSExceptions which have different properties, I'm asking if it's possible to somehow get at the original managed exception's properties.

**Comment by chamons** [comment]
Sorry I misread (what I get for doing work without sufficient ☕️ ).

I believe the relevant code here here: https://github.com/xamarin/xamarin-macios/blob/master/runtime/runtime.m#L2337 

It appears we are pulling out the name / message and shoving the exception itself into the userInfo.

Given sufficient hacking, I think you can get what you need.

**Comment by lemonmojo** [comment]
Thx for the pointers, chamons! I guess I can hack something together with the additional info. =)
